### PR TITLE
Update: auto-fix for `comma-style`  (fixes #6941)

### LINKED ISSUE #6941: --fix isn't handling "',' should be placed last  comma-style" errors
**What version of ESLint are you using?** - v3.3.1

**What parser (default, Babel-ESLint, etc.) are you using?** - babel-eslint

**Please show your full configuration:**

```json
{
  "parser": "babel-eslint",
  "extends": "standard",
  "plugins": [
    "flowtype"
  ],
  "rules": {
      "flowtype/define-flow-type": 1,
      "flowtype/space-before-type-colon": [
          1,
          "never"
      ],
      "flowtype/use-flow-type": 1,
      "flowtype/valid-syntax": 1,
      "no-duplicate-imports": 0
  },
  "settings": {
      "flowtype": {
          "onlyFilesWithFlowAnnotation": false
      }
  }
}
```

Standar is defined here https://github.com/feross/eslint-config-standard/blob/master/eslintrc.json

**What did you do? Please include the actual source code causing the issue.**

```js
export var obj =
  { a: 1
  , b: 2
  }
```

**What did you expect to happen?** 

ESLint should fix comma style by placing at the end:

```js
export var obj =
  { a: 1,
    b: 2
  }
```

**What actually happened? Please include the actual, raw output from ESLint.**

```sh
eslint --fix ./example.js

/Users/gozala/Projects/browser.html/example.js
  3:4  error  ',' should be placed last  comma-style
```

### PR DESCRIPTION
**Pull request by gyandeeps**
**What issue does this pull request address?**

Introduce auto-fix for `comma-style` rule.

**What changes did you make? (Give an overview)**

Fix the position of comma.

**Is there anything you'd like reviewers to focus on?**

This is just a proof of concept PR. 

**UPDATE:**

Its ready for review and merge.

### CODE REVIEW

File: lib/rules/comma-style.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ module.exports = {
             category: "Stylistic Issues",
             recommended: false
         },
-
+        fixable: true,
```
Comment (platinumazure) [inline lib/rules/comma-style.js]: I think we usually want "code" or "whitespace" here. Not sure which this is though!

Comment (ilyavolodin) [inline lib/rules/comma-style.js]: This would be "code". "whitespace" is basically just for spacing fixes.

File: lib/rules/comma-style.js [hunk 2]
```diff
@@ -82,21 +82,50 @@ module.exports = {
...
+                    loc: {
+                        line: commaToken.loc.end.line,
+                        column: commaToken.loc.start.column
+                    },
+                    message: "Bad line breaking before and after ','.",
+                    fix(fixer) {
+                        return fixer.removeRange([
+                            previousItemToken.range[1],
+                            commaToken.range[0]
+                        ]);
```
Comment (mysticatea) [inline lib/rules/comma-style.js]: This fixing seems to be for `style === "last"`.
If `style === "first"`, probably it should be `fixer.removeRange([commaToken.range[1], currentItemToken.range[0]])`.

**Comment by mysticatea** [comment]
I think this rule should not remove comments.
We often write comments between properties.

```js
let obj = {
    foo: 1,  // a comment
    bar: 2,  // another comment

    /**
     * a doc comment
     */
    method: function() {
    },
}
```

**Comment by gyandeeps** [comment]
As always @mysticatea to the rescue. Excellent idea. So far looks good.

File: lib/rules/comma-style.js [hunk 3]
```diff
@@ -62,6 +62,49 @@ module.exports = {
...
+         * Modified text based on the style
+         * @param {string} styleType Style type
+         * @param {string} text Source code text
+         * @returns {string} modified text
+         * @private
+         */
+        function getReplacedText(styleType, text) {
+            switch (styleType) {
+                case "between":
+                    return `,${text.replace("\n", "")}`;
```
Comment (mysticatea) [inline lib/rules/comma-style.js]: This is dangerous.

```js
let obj = {
    a // a comment
    ,
    b
}
```

I think good to handle the same way as others for lonely comma's cases also. I.e. `getFixerFunction(style, previousItemToken, commaToken, currentItemToken)`

Or maybe it should not fix the lonely commas.

Comment (gyandeeps) [inline lib/rules/comma-style.js]: @mysticatea I din't quite get what you trying to say here.

I have a unit test covering the above use case, and the fixed code would be 

```js
let obj = {
    a, // a comment
    b
}
```

Comment (mysticatea) [inline lib/rules/comma-style.js]: Oops, sorry, it was my misread.
I sometimes forget that `String.prototype.replace` replaces only the first matched part.

**Comment by mysticatea** [comment]
LGTM

**Comment by nzakas** [comment]
Just want to double check that we are only doing fixes that are 100% safe with this now?

**Comment by mysticatea** [comment]
@nzakas I think so. The logic is:

```
let obj = { foo: 1 // comment\n  /**\n   * comment\n   */\n,   bar: 2 }
                  └┬──────────────────────────────────────────┘
                   └ Part A - between the end of the previous property and the current property.

let obj = { foo: 1 // comment\n  /**\n   * comment\n   */\n,   bar: 2 }
                  └┬──────────────────────────────────────┘
                   └ Part B - between the end of the previous property and the comma.

let obj = { foo: 1 // comment\n  /**\n   * comment\n   */\n,   bar: 2 }
                                                            └┬┘
        Part C - between the comma and the current property. ┘
```

Then, it replaces `A` by `"," + B + C` or `B + C + ","`.
It will move only the comma exactly.

In addition, in `"first"` cases, `comma-spacing` and `indent` rules will tweak the code in the next passes of autofix.

**Comment by ilyavolodin** [comment]
I think this is ready to be merged, but I'm not 100% if all of the comments have been resolved. @mysticatea could you confirm and merge if you think it's good to go?

**Comment by mysticatea** [comment]
Sure, LGTM.

**Pull request merged**